### ISSUE TITLE
Retries to KeyMgmt node not working properly

### ISSUE THREAD

**Issue by lacabra** [enigmampc/enigma-p2p#187]
**Describe the bug**
After merging Core's [Issue 121](https://github.com/enigmampc/enigma-core/pull/121), the Principal Node waits to verify that the workerParams have been set properly to respond to a `GetStateKeys` request from `p2p`. Yet, `p2p` makes that request immediately upon the contract emitting the event that the parameters have been set, and before the KeyMgmt node has had time to verify it. 

The connection succeeds, but there is an error in the response, and so the retries do not kick in.

**To Reproduce**
1. Launch the `integration-tests` network and deploy a contract

The deployment will succeed, but look at the network output, and you will see the p2p printing a backtrace from the principal node that should not be there.

**Expected behavior**
The `p2p` should retry if there is an error in the response, or shouldn't ? (cc: @fredfortier)

**Backtrace**
```
p2p_1        | [Mon May 06 2019 22:18:24 GMT+0000 (Coordinated Universal Time)] DEBUG Connecting to principal node: http://172.25.0.4:3040
p2p_1        | [Mon May 06 2019 22:18:24 GMT+0000 (Coordinated Universal Time)] ERROR Failed Principal node connection: -32603 - Unable to get keys: ErrorMessage { msg: "Cannot get contract addresses until the EpochState is confirmed." }
p2p_1        | 
p2p_1        | stack backtrace:
p2p_1        |    0: failure::backtrace::internal::InternalBacktrace::new::h95effe03f49df865 (0x557eb9415e1e)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.5/src/backtrace/internal.rs:44
p2p_1        |    1: failure::backtrace::Backtrace::new::hd5aa352cdf4f33fc (0x557eb94158ed)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.5/src/backtrace/mod.rs:111
p2p_1        |    2: <failure::error::error_impl::ErrorImpl as core::convert::From<F>>::from::h80b1411ba7e6cf35 (0x557eb84b1eb7)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.5/src/error/error_impl.rs:19
p2p_1        |    3: <failure::error::Error as core::convert::From<F>>::from::h837a41c7d56f73df (0x557eb8561e7c)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.5/src/error/mod.rs:36
p2p_1        |    4: failure::error_message::err_msg::h9af1501399ab3ad6 (0x557eb85989b0)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.5/src/error_message.rs:12
p2p_1        |    5: enigma_principal_app::epoch_u::epoch_types::EpochState::get_contract_addresses::{{closure}}::h1759e3cc81ceb158 (0x557eb857086d)
p2p_1        |              at src/epoch_u/epoch_types.rs:78
p2p_1        |    6: enigma_principal_app::epoch_u::epoch_types::EpochState::get_contract_addresses::h825e72130e9c5efc (0x557eb84e3231)
p2p_1        |              at src/epoch_u/epoch_types.rs:66
p2p_1        |    7: enigma_principal_app::boot_network::keys_provider_http::PrincipalHttpServer::find_epoch_contract_addresses::{{closure}}::h120b279852f89d76 (0x557eb85d8214)
p2p_1        |              at src/boot_network/keys_provider_http.rs:85
p2p_1        |    8: enigma_principal_app::boot_network::keys_provider_http::PrincipalHttpServer::find_epoch_contract_addresses::h1fc1cb9883b8f4c8 (0x557eb859a119)
p2p_1        |              at src/boot_network/keys_provider_http.rs:79
p2p_1        |    9: enigma_principal_app::boot_network::keys_provider_http::PrincipalHttpServer::get_state_keys::{{closure}}::h8d6b8fb45cfb8d22 (0x557eb85d98eb)
p2p_1        |              at src/boot_network/keys_provider_http.rs:101
p2p_1        |   10: enigma_principal_app::boot_network::keys_provider_http::PrincipalHttpServer::get_state_keys::hd5b2725a94bd8298 (0x557eb859a76f)
p2p_1        |              at src/boot_network/keys_provider_http.rs:89
p2p_1        |   11: enigma_principal_app::boot_network::keys_provider_http::PrincipalHttpServer::start::{{closure}}::{{closure}}::h232512e1938c1abc (0x557eb85da38c)
p2p_1        |              at src/boot_network/keys_provider_http.rs:121
p2p_1        |   12: <F as jsonrpc_core::calls::RpcMethodSimple>::call::h19f92912b97a4347 (0x557eb85d4227)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/jsonrpc-core-11.0.0/src/calls.rs:70
p2p_1        |   13: <jsonrpc_core::io::MetaIoHandler<T, S>>::add_method::{{closure}}::h5764247f6e8791bb (0x557eb8562fe0)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/jsonrpc-core-11.0.0/src/io.rs:126
p2p_1        |   14: <F as jsonrpc_core::calls::RpcMethod<T>>::call::h57c0521cf43a3cce (0x557eb8565701)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/jsonrpc-core-11.0.0/src/calls.rs:91
p2p_1        |   15: <jsonrpc_core::io::MetaIoHandler<T, S>>::handle_call::{{closure}}::{{closure}}::{{closure}}::h360b66a30ca86b20 (0x557eb85647d3)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/jsonrpc-core-11.0.0/src/io.rs:243
p2p_1        |   16: <futures::future::lazy::Lazy<F, R>>::get::hce828a2426f7e9c5 (0x557eb848dd4f)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/lazy.rs:64
p2p_1        |   17: <futures::future::lazy::Lazy<F, R> as futures::future::Future>::poll::he40677257e2d2b30 (0x557eb848e102)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/lazy.rs:82
p2p_1        |   18: <futures::future::chain::Chain<A, B, C>>::poll::hb2618029b6c73570 (0x557eb857226a)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/chain.rs:26
p2p_1        |   19: <futures::future::then::Then<A, B, F> as futures::future::Future>::poll::ha141415aeefe1068 (0x557eb849a079)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/then.rs:32
p2p_1        |   20: <alloc::boxed::Box<F> as futures::future::Future>::poll::h83ca15d531311539 (0x557eb860387b)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/mod.rs:113
p2p_1        |   21: <futures::future::either::Either<A, B> as futures::future::Future>::poll::hd96fbc096d14f1e1 (0x557eb8599289)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/either.rs:35
p2p_1        |   22: <futures::future::either::Either<A, B> as futures::future::Future>::poll::h326a0d2ff00d0fc1 (0x557eb85990f8)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/either.rs:36
p2p_1        |   23: <futures::future::map::Map<A, F> as futures::future::Future>::poll::h604a071b71d04944 (0x557eb8567f6d)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/map.rs:30
p2p_1        |   24: <futures::future::either::Either<A, B> as futures::future::Future>::poll::he9d32f0853da786f (0x557eb8599319)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/either.rs:35
p2p_1        |   25: <futures::future::either::Either<A, B> as futures::future::Future>::poll::hd4c2ecbeef5e5008 (0x557eb8599218)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/either.rs:36
p2p_1        |   26: <futures::future::either::Either<A, B> as futures::future::Future>::poll::h70d14be6d5528ea2 (0x557eb8599188)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/either.rs:36
p2p_1        |   27: <futures::future::map::Map<A, F> as futures::future::Future>::poll::h09a5f6e59623c999 (0x557eb85676cd)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/map.rs:30
p2p_1        |   28: <jsonrpc_http_server::handler::RpcHandler<M, S> as futures::future::Future>::poll::h65960c16a0b316e8 (0x557eb85db501)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/jsonrpc-http-server-11.0.0/src/handler.rs:279
p2p_1        |   29: <jsonrpc_http_server::handler::RpcHandler<M, S> as futures::future::Future>::poll::h65960c16a0b316e8 (0x557eb85dcd7d)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/jsonrpc-http-server-11.0.0/src/handler.rs:317
p2p_1        |   30: <jsonrpc_http_server::handler::RpcHandler<M, S> as futures::future::Future>::poll::h65960c16a0b316e8 (0x557eb85dcd7d)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/jsonrpc-http-server-11.0.0/src/handler.rs:317
p2p_1        |   31: <jsonrpc_http_server::handler::Handler<M, S> as futures::future::Future>::poll::hb9d7f0874daff8b1 (0x557eb85e336f)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/jsonrpc-http-server-11.0.0/src/handler.rs:126
p2p_1        |   32: <hyper::proto::h1::dispatch::Server<S> as hyper::proto::h1::dispatch::Dispatch>::poll_msg::hb6eeaf19e2c96e76 (0x557eb8589ed8)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.12.28/src/proto/h1/dispatch.rs:378
p2p_1        |   33: <hyper::proto::h1::dispatch::Dispatcher<D, Bs, I, T>>::poll_write::h6fc797169e05d4b9 (0x557eb858c875)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.12.28/src/proto/h1/dispatch.rs:236
p2p_1        |   34: <hyper::proto::h1::dispatch::Dispatcher<D, Bs, I, T>>::poll_inner::hf42ad1bd45eedef8 (0x557eb858bcd3)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.12.28/src/proto/h1/dispatch.rs:101
p2p_1        |   35: <hyper::proto::h1::dispatch::Dispatcher<D, Bs, I, T>>::poll_catch::h0a954efba49b6a86 (0x557eb858b738)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.12.28/src/proto/h1/dispatch.rs:86
p2p_1        |   36: <hyper::proto::h1::dispatch::Dispatcher<D, Bs, I, T> as futures::future::Future>::poll::h3c3db11cd0d325e0 (0x557eb858b39f)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.12.28/src/proto/h1/dispatch.rs:347
p2p_1        |   37: <futures::future::either::Either<A, B> as futures::future::Future>::poll::h2ee1341df2b4301f (0x557eb8599049)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/either.rs:35
p2p_1        |   38: futures::future::option::<impl futures::future::Future for core::option::Option<F>>::poll::h64477f7db2a87ca7 (0x557eb85ba191)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/option.rs:12
p2p_1        |   39: <hyper::server::conn::Connection<I, S, E> as futures::future::Future>::poll::h3b75c28bb97f7a1d (0x557eb8588493)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.12.28/src/server/conn.rs:635
p2p_1        |   40: <futures::future::map_err::MapErr<A, F> as futures::future::Future>::poll::h25231581e417d4b9 (0x557eb857e94c)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/map_err.rs:30
p2p_1        |   41: <alloc::boxed::Box<F> as futures::future::Future>::poll::h4595b1635045551e (0x557eb8f93774)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/future/mod.rs:113
p2p_1        |   42: <futures::task_impl::Spawn<T>>::poll_future_notify::{{closure}}::h35ae98f57af0bb80 (0x557eb8d88ed2)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/task_impl/mod.rs:329
p2p_1        |   43: <futures::task_impl::Spawn<T>>::enter::{{closure}}::h36268d5a36e6addb (0x557eb8d8918f)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/task_impl/mod.rs:399
p2p_1        |   44: futures::task_impl::std::set::hb6067b4529d35696 (0x557eb8db2f9b)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/task_impl/std/mod.rs:78
p2p_1        |   45: <futures::task_impl::Spawn<T>>::enter::hc928d4dfe8cba8ad (0x557eb8d88fda)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/task_impl/mod.rs:399
p2p_1        |   46: <futures::task_impl::Spawn<T>>::poll_fn_notify::hf2e31a90791515c0 (0x557eb8d88d73)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/task_impl/mod.rs:291
p2p_1        |   47: <futures::task_impl::Spawn<T>>::poll_future_notify::hd60670d3bc6801b9 (0x557eb8d88ea6)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.26/src/task_impl/mod.rs:329
p2p_1        |   48: tokio_threadpool::task::Task::run::{{closure}}::hfc09da49b1cccf69 (0x557eb8d8c514)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.14/src/task/mod.rs:145
p2p_1        |   49: core::ops::function::FnOnce::call_once::h1aa3aa12500fb965 (0x557eb8da29d8)
p2p_1        |              at libcore/ops/function.rs:238
p2p_1        |   50: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hf85fc5f5057e9933 (0x557eb8d8cbf7)
p2p_1        |              at libstd/panic.rs:313
p2p_1        |   51: std::panicking::try::do_call::h75daf33d47c260ea (0x557eb8d88bc0)
p2p_1        |              at libstd/panicking.rs:310
p2p_1        |   52: __rust_maybe_catch_panic (0x557eb950a4b9)
p2p_1        |              at libpanic_unwind/lib.rs:102
p2p_1        |   53: std::panicking::try::h46f709c0df486dbd (0x557eb8d8897e)
p2p_1        |              at libstd/panicking.rs:289
p2p_1        |   54: std::panic::catch_unwind::hfe30c38a722f4939 (0x557eb8d8cc8f)
p2p_1        |              at libstd/panic.rs:392
p2p_1        |   55: tokio_threadpool::task::Task::run::hf132bfbe3ee8344e (0x557eb8d8bd38)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.14/src/task/mod.rs:130
p2p_1        |   56: tokio_threadpool::worker::Worker::run_task2::h15116d76271172e0 (0x557eb8da7b3a)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.14/src/worker/mod.rs:567
p2p_1        |   57: tokio_threadpool::worker::Worker::run_task::h45f92818975a1325 (0x557eb8da743f)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.14/src/worker/mod.rs:459
p2p_1        |   58: tokio_threadpool::worker::Worker::try_run_owned_task::h9ae4b82b054c6c1d (0x557eb8da6ae9)
principal_1  | Found selected worker: 0xb6b1ef8f966183bebf88fda6e1b18ef9a024e68d for contract: Hash256([40, 34, 216, 151, 113, 173, 46, 55, 95, 202, 210, 3, 64, 144, 28, 113, 105, 37, 195, 67, 164, 142, 130, 168, 100, 209, 114, 116, 19, 111, 138, 37])
principal_1  | First selected worker inserted
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.14/src/worker/mod.rs:390
p2p_1        |   59: tokio_threadpool::worker::Worker::try_run_task::h05f779e067b25c7b (0x557eb8da64ec)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.14/src/worker/mod.rs:297
p2p_1        |   60: tokio_threadpool::worker::Worker::run::hfe50d40622e0fa37 (0x557eb8da635b)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.14/src/worker/mod.rs:241
p2p_1        |   61: tokio::runtime::threadpool::builder::Builder::build::{{closure}}::{{closure}}::{{closure}}::{{closure}}::{{closure}}::he413f56310a921bf (0x557eb8d13100)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.19/src/runtime/threadpool/builder.rs:349
p2p_1        |   62: tokio_trace_core::dispatcher::with_default::h71d4c687a502abdc (0x557eb8d2a7d2)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-trace-core-0.1.0/src/dispatcher.rs:54
p2p_1        |   63: tokio::runtime::threadpool::builder::Builder::build::{{closure}}::{{closure}}::{{closure}}::{{closure}}::h9c8773d8edf9ae21 (0x557eb8d1313b)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.19/src/runtime/threadpool/builder.rs:348
p2p_1        |   64: tokio_timer::timer::handle::with_default::{{closure}}::h00f0bdb818c779e4 (0x557eb8d14b73)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.10/src/timer/handle.rs:94
p2p_1        |   65: <std::thread::local::LocalKey<T>>::try_with::he5c9e85e6b613177 (0x557eb8d305bd)
p2p_1        |              at libstd/thread/local.rs:294
p2p_1        |   66: <std::thread::local::LocalKey<T>>::with::hb9ac9a0956b540b1 (0x557eb8d2f1ca)
p2p_1        |              at libstd/thread/local.rs:248
p2p_1        |   67: tokio_timer::timer::handle::with_default::hd81adfb11e0cf2ec (0x557eb8d149cd)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.10/src/timer/handle.rs:81
p2p_1        |   68: tokio::runtime::threadpool::builder::Builder::build::{{closure}}::{{closure}}::{{closure}}::h83649487ab1c284b (0x557eb8d131a5)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.19/src/runtime/threadpool/builder.rs:347
p2p_1        |   69: tokio_timer::clock::clock::with_default::{{closure}}::hd4d29d8e72965e91 (0x557eb8d26e89)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.10/src/clock/clock.rs:141
p2p_1        |   70: <std::thread::local::LocalKey<T>>::try_with::heab962244234e85a (0x557eb8d30822)
p2p_1        |              at libstd/thread/local.rs:294
p2p_1        |   71: <std::thread::local::LocalKey<T>>::with::h7a5b68ec10d54c1d (0x557eb8d2f10d)
p2p_1        |              at libstd/thread/local.rs:248
p2p_1        |   72: tokio_timer::clock::clock::with_default::he9b733acd220f610 (0x557eb8d26d88)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.10/src/clock/clock.rs:124
p2p_1        |   73: tokio::runtime::threadpool::builder::Builder::build::{{closure}}::{{closure}}::h33b82a475f946842 (0x557eb8d131f2)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.19/src/runtime/threadpool/builder.rs:346
p2p_1        |   74: tokio_reactor::with_default::{{closure}}::hbf139a04a2c078e7 (0x557eb8d14f19)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.9/src/lib.rs:237
p2p_1        |   75: <std::thread::local::LocalKey<T>>::try_with::h26979826afa2f401 (0x557eb8d2f455)
p2p_1        |              at libstd/thread/local.rs:294
p2p_1        |   76: <std::thread::local::LocalKey<T>>::with::h2fb7ad9e54e3327b (0x557eb8d2f04d)
p2p_1        |              at libstd/thread/local.rs:248
p2p_1        |   77: tokio_reactor::with_default::h2c1e25bc9a4a596e (0x557eb8d14cec)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.9/src/lib.rs:217
p2p_1        |   78: tokio::runtime::threadpool::builder::Builder::build::{{closure}}::haf01b691685c37ac (0x557eb8d13297)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.19/src/runtime/threadpool/builder.rs:345
p2p_1        |   79: tokio_threadpool::callback::Callback::call::h73b5e1e35edbd129 (0x557eb8d91501)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.14/src/callback.rs:22
p2p_1        |   80: tokio_threadpool::worker::Worker::do_run::{{closure}}::{{closure}}::hf0e22e9544367ed7 (0x557eb8da601e)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.14/src/worker/mod.rs:127
p2p_1        |   81: tokio_executor::global::with_default::{{closure}}::ha91cc700fbf3fec5 (0x557eb8db0bf9)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.7/src/global.rs:209
p2p_1        |   82: <std::thread::local::LocalKey<T>>::try_with::h673bc3ad38b6a3e5 (0x557eb8da0806)
p2p_1        |              at libstd/thread/local.rs:294
p2p_1        |   83: <std::thread::local::LocalKey<T>>::with::ha8a68b1b9c04947a (0x557eb8d9fed1)
p2p_1        |              at libstd/thread/local.rs:248
p2p_1        |   84: tokio_executor::global::with_default::hfa94ef8aa57ab9f0 (0x557eb8db0ae4)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.7/src/global.rs:178
p2p_1        |   85: tokio_threadpool::worker::Worker::do_run::{{closure}}::h4cf9c44555b597a7 (0x557eb8da60bd)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.14/src/worker/mod.rs:125
p2p_1        |   86: <std::thread::local::LocalKey<T>>::try_with::h76e4f7e74c79e761 (0x557eb8da0a44)
p2p_1        |              at libstd/thread/local.rs:294
p2p_1        |   87: <std::thread::local::LocalKey<T>>::with::h3bb4a4e73726634f (0x557eb8d9fe0c)
p2p_1        |              at libstd/thread/local.rs:248
p2p_1        |   88: tokio_threadpool::worker::Worker::do_run::h2a7832e7e5b726cf (0x557eb8da5ef6)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.14/src/worker/mod.rs:116
p2p_1        |   89: tokio_threadpool::pool::Pool::spawn_thread::{{closure}}::hb186409f852dc373 (0x557eb8dac04a)
p2p_1        |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.14/src/pool/mod.rs:344
p2p_1        |   90: std::sys_common::backtrace::__rust_begin_short_backtrace::hc2d3f7cbbaf16bba (0x557eb8db6fcd)
p2p_1        |              at libstd/sys_common/backtrace.rs:136
p2p_1        |   91: std::thread::Builder::spawn::{{closure}}::{{closure}}::hfedded1b22c644c4 (0x557eb8d95a80)
p2p_1        |              at libstd/thread/mod.rs:409
p2p_1        |   92: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::ha54c7075806bb500 (0x557eb8d8cbc0)
p2p_1        |              at libstd/panic.rs:313
p2p_1        |   93: std::panicking::try::do_call::h85432378fecb11c1 (0x557eb8d88c49)
p2p_1        |              at libstd/panicking.rs:310
p2p_1        |   94: __rust_maybe_catch_panic (0x557eb950a4b9)
p2p_1        |              at libpanic_unwind/lib.rs:102
p2p_1        |   95: std::panicking::try::h6461ec30123372f6 (0x557eb8d88ab4)
p2p_1        |              at libstd/panicking.rs:289
p2p_1        |   96: std::panic::catch_unwind::h38b65e3762aac62c (0x557eb8d8cc40)
p2p_1        |              at libstd/panic.rs:392
p2p_1        |   97: std::thread::Builder::spawn::{{closure}}::hf53b6233a3d4ea35 (0x557eb8d95888)
p2p_1        |              at libstd/thread/mod.rs:408
p2p_1        |   98: <F as alloc::boxed::FnBox<A>>::call_box::h6c19396be78441ae (0x557eb8d9866f)
p2p_1        |              at liballoc/boxed.rs:661
p2p_1        |   99: <alloc::boxed::Box<(dyn alloc::boxed::FnBox<A, Output=R> + 'a)> as core::ops::function::FnOnce<A>>::call_once::h0ef3831196409bcd (0x557eb94fc61a)
p2p_1        |              at liballoc/boxed.rs:671
p2p_1        |       std::sys_common::thread::start_thread::hec6cb70358a5b33c
p2p_1        |              at libstd/sys_common/thread.rs:24
p2p_1        |  100: std::sys::unix::thread::Thread::new::thread_start::hfaaf486d6bd1eb3f (0x557eb94e8735)
p2p_1        |              at libstd/sys/unix/thread.rs:90
p2p_1        |  101: start_thread (0x7fe8f2fcf6da)
p2p_1        |  102: __clone (0x7fe8f2ae088e)
p2p_1        |  103: <unknown> (0x0)
```

**Additional context**
I was expecting that changing [this line](https://github.com/enigmampc/enigma-p2p/blob/4808fe91de3e0f2e6e4bce0ff4a5cc4c215b6052/src/worker/handlers/PrincipalNode.js#L33) to the following would have addressed the issue, but it didn't seem to make a difference, why is that?
```
if ((err || response.error) && operation.retry(err)) return; 
```

**Comment by lenak25** [comment]
Hi @lacabra ,

Yes, I would also expect that the change you've suggested would work. Any chance that the error was a result of reaching the retries number, once you made the change?
I would also expect the KM node not to send its backtrace to the worker... @fredfortier

**Comment by lacabra** [comment]
The error was thrown immediately, so I'm positive the retries number was not reached. As far as I could see there was no retry at all. Will investigate more and report back.

**Comment by lacabra** [comment]
Fixed via #188 (merged to #209) and #211.
